src/trajectory.jl

```diff
@@ -664,115 +721,311 @@ function check_right_subtree(
+    sampler′ = combine(rng, state.sampler, state′.sampler)
+    termination′ = state.termination * state′.termination * isterminated(h, tree′, treeleft, treeright)
+    state′ = TreeState(tree′, sampler′, termination′)
+    is_accept = .!isterminated(state)
+    state′ = accept!(state, state′, is_accept)
+    return state′
+end
+
+"""
+Combine `state` and `state′` with candidate point `zcand` for all chains where `state` has
+not terminated.
+"""
+function combine(zcand::PhasePoint, h::Hamiltonian, state::TreeState, state′::TreeState, v)
+    treeleft, treeright = left_right_subtrees(state.tree, state′.tree, v)
+    tree′ = combine(treeleft, treeright)
+    sampler′ = combine(zcand, state.sampler, state′.sampler)
+    termination′ = state.termination * state′.termination * isterminated(h, tree′, treeleft, treeright)
+    state′ = TreeState(tree′, sampler′, termination′)
+    is_accept = .!isterminated(state)
+    state′ = accept!(state, state′, is_accept)
+    return state′
+end
+
+"""
+Create a level 0 `TreeState` at the initial point `z0`.
+"""
+function init_tree_state(rng::AbstractRNG, τ::NUTS{S,C}, z0::PhasePoint) where {S,C}
+    H0 = energy(z0)
+    tree = BinaryTree(z0, z0, C(z0), zero(H0), map(_ -> 0, H0), zero(H0))
+    sampler = S(rng, z0)
+    # to get same type of array as boolean check
+    term = H0 .> Inf
+    termination = Termination(term, deepcopy(term))
+    state = TreeState(tree, sampler, termination)
+    return state
+end
+
+function build_one_leaf_tree(nt::NUTS{S,C}, h, z, sampler, v, H0) where {S,C}
+    z′ = step(nt.integrator, h, z, 1; fwd = v .> 0)
+    H′ = energy(z′)
+    ΔH = H′ - H0
+    α′ = exp.(min.(0, .-ΔH))
+    tree′ = BinaryTree(z′, z′, C(z′), α′, map(_ -> 1, v), ΔH)
+    sampler′ = S(sampler, H0, z′)
+    termination′ = Termination(sampler′, nt, H0, H′)
+    return z′, TreeState(tree′, sampler′, termination′)
+end
+
+"""
+    subtree_cache_combine_range(ileaf::Integer) -> UnitRange
+
+Get the indices of the cached trees that need to be combined with the two-leaf tree produced
+at even leaf number `ileaf`.
+"""
+@inline function subtree_cache_combine_range(ileaf)
+    imin = count_ones(ileaf)
+    inum = trailing_zeros(ileaf)
+    return (imin + inum - 1):-1:imin
+end
+
+function tree_extension_phasepoint(tree, v::Int)
+    return v == -1 ? tree.zleft : tree.zright
+end
+function tree_extension_phasepoint(tree, v)
+    isright = isone.(v)
+    z′ = deepcopy(tree.zright)
+    z′ = accept!(tree.zleft, z′, isright)
+    return z′
+end
+
+function left_right_subtrees(tree, tree′, v::Int)
+    return v == -1 ? (tree′, tree) : (tree, tree′)
+end
+function left_right_subtrees(tree, tree′, v)
+    tree_left, tree_right = deepcopy(tree), deepcopy(tree′)
+    is_accept = isone.(v)
+    tree_right = accept!(tree, tree_right, is_accept)
+    tree_left = accept!(tree′, tree_left, is_accept)
+    return tree_left, tree_right
+end
+
 """
 Recursivly build a tree for a given depth `j`.
 """
 function build_tree(
```

Comment:
I realise after doing this refactoring to the existing NUTS, we cannot check if the new implementation gives the same results as the old one. Any idea on doing this?